From 4e37bef8480dcb559a893a3c32a1194322ba19b3 Mon Sep 17 00:00:00 2001
From: Kevin Chen <kevin_chen@aspeedtech.com>
Date: Thu, 21 Aug 2025 17:23:12 +0800
Subject: [PATCH 12/32] misc: edaf_bridge: Add eDAF bridge configuration for
 erase support

Add support for configuring the eDAF bridge to enable erase functionality
based on the platform. Introduce new register definitions and extend the
probe function to handle the "edaf-earse-with-1" property. Map the eDAF
bridge configuration registers and update the global configuration register
accordingly.

Change-Id: Iebf83fec3fc77ace1326c2e98fc5fe069c5ce1fc
Signed-off-by: Kevin Chen <kevin_chen@aspeedtech.com>
(cherry picked from commit 0be6a527fc497e3241c658aaf3f4b8174c320054)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/misc/aspeed_edaf_bridge.c | 127 ++++++++++++++++++++----------
 1 file changed, 84 insertions(+), 43 deletions(-)

diff --git a/drivers/misc/aspeed_edaf_bridge.c b/drivers/misc/aspeed_edaf_bridge.c
index d26f69115c..7fa02b84ad 100644
--- a/drivers/misc/aspeed_edaf_bridge.c
+++ b/drivers/misc/aspeed_edaf_bridge.c
@@ -40,6 +40,10 @@
 #define   EDAF_BDGE_MISC_MBASE_H	GENMASK(15, 8)
 #define   EDAF_BDGE_MISC_CBASE_H	GENMASK(7, 0)
 
+#define EDAF_BDGE_GLOBAL_CFG		0x0
+#define   EDAF_BDGE_GLOBAL_CFG0_EARSE	BIT(1)
+#define   EDAF_BDGE_GLOBAL_CFG1_EARSE	BIT(9)
+
 uint64_t ast27xx_soc_virt_addr_to_phy_addr(uintptr_t addr)
 {
 	if (addr < 0x80000000)
@@ -58,11 +62,11 @@ uintptr_t ast27xx_soc_phy_addr_to_virt_addr(uint64_t addr)
 
 static int aspeed_edaf_bridge_probe(struct udevice *dev)
 {
-	void *edaf_bridge_regs;
+	void *edaf_bridge_regs, *edaf_bridge_cfg_regs;
 	void *scu_regs, *espi_regs, *virt;
-	ofnode node, scu1_node, mem_node;
+	ofnode node, scu1_node, mem_node, edaf_bridge_cfg_node;
 	uint64_t cbase, mbase, espibase;
-	uint32_t scu, edaf, espi, phandle;
+	uint32_t scu, edaf, edaf_ddr_mode, edaf_gcfg, espi, phandle;
 	uint32_t plat = (unsigned long)dev_get_driver_data(dev);
 	struct resource res;
 	int rc;
@@ -118,8 +122,17 @@ static int aspeed_edaf_bridge_probe(struct udevice *dev)
 		return -EINVAL;
 	}
 
+	edaf_ddr_mode = ofnode_read_bool(node, "edaf-ddr-mode");
+
 	edaf = readl(edaf_bridge_regs + EDAF_BDGE_CFG);
-	if (ofnode_read_bool(node, "edaf-ddr-mode")) {
+	if (edaf_ddr_mode)
+		edaf &= ~EDAF_BDGE_CFG_CMD_EN;
+	else
+		edaf |= EDAF_BDGE_CFG_CMD_EN;
+	writel(edaf, edaf_bridge_regs + EDAF_BDGE_CFG);
+
+	if (edaf_ddr_mode) {
+		/* eDAF on reserved memory via DDR mode */
 		rc = ofnode_read_u32(node, "mem-base", &phandle);
 		if (rc) {
 			printf("cannot get mem-base phandle\n");
@@ -145,59 +158,87 @@ static int aspeed_edaf_bridge_probe(struct udevice *dev)
 			return -ENOMEM;
 		}
 
-		edaf &= ~EDAF_BDGE_CFG_CMD_EN;
+		if (ofnode_read_bool(node, "edaf-earse-with-1")) {
+			rc = ofnode_read_u32(node, "aspeed,edaf_bridge_cfg", &phandle);
+			if (rc) {
+				printf("cannot get edaf_bridge_cfg phandle\n");
+				return -ENODEV;
+			}
+
+			edaf_bridge_cfg_node = ofnode_get_by_phandle(phandle);
+			if (!ofnode_valid(edaf_bridge_cfg_node)) {
+				printf("cannot get eDAF bridge GCFG device node\n");
+				return -ENODEV;
+			}
+
+			edaf_bridge_cfg_regs = (void *)ofnode_get_addr(edaf_bridge_cfg_node);
+			if (edaf_bridge_cfg_regs == (void *)FDT_ADDR_T_NONE) {
+				printf("cannot map eDAF bridge GCFG registers\n");
+				return -ENODEV;
+			}
+
+			edaf_gcfg = readl(edaf_bridge_cfg_regs + EDAF_BDGE_GLOBAL_CFG);
+			if (plat == EDAF_BDGE_OVER_ESPI0) {
+				edaf_gcfg |= EDAF_BDGE_GLOBAL_CFG0_EARSE;
+			} else if (plat == EDAF_BDGE_OVER_ESPI1) {
+				edaf_gcfg |= EDAF_BDGE_GLOBAL_CFG1_EARSE;
+			} else {
+				printf("Unknown platform for eDAF bridge\n");
+				return -ENODEV;
+			}
+			writel(edaf_gcfg, edaf_bridge_cfg_regs + EDAF_BDGE_GLOBAL_CFG);
+		}
+
 	} else {
+		/* eDAF on SPI1 or SPI2 */
 		rc = ofnode_read_u64(node, "mem-base", &mbase);
 		if (rc)
 			printf("cannot get eDAF memory region\n");
 
-		edaf |= EDAF_BDGE_CFG_CMD_EN;
-	}
-	writel(mbase, edaf_bridge_regs + EDAF_BDGE_MBASE);
-	writel(edaf, edaf_bridge_regs + EDAF_BDGE_CFG);
+		rc = ofnode_read_u64(node, "ctl-base", &cbase);
+		if (!rc) {
+			writel(cbase, edaf_bridge_regs + EDAF_BDGE_CBASE);
 
-	edaf = readl(edaf_bridge_regs + EDAF_BDGE_MISC);
-	edaf &= ~EDAF_BDGE_MISC_MBASE_H;
-	edaf |= FIELD_PREP(EDAF_BDGE_MISC_MBASE_H, mbase >> 32);
-	writel(edaf, edaf_bridge_regs + EDAF_BDGE_MISC);
-
-	rc = ofnode_read_u64(node, "ctl-base", &cbase);
-	if (!rc) {
-		writel(cbase, edaf_bridge_regs + EDAF_BDGE_CBASE);
+			edaf = readl(edaf_bridge_regs + EDAF_BDGE_MISC);
+			edaf &= ~EDAF_BDGE_MISC_CBASE_H;
+			edaf |= FIELD_PREP(EDAF_BDGE_MISC_CBASE_H, cbase >> 32);
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_MISC);
+		}
 
-		edaf = readl(edaf_bridge_regs + EDAF_BDGE_MISC);
-		edaf &= ~EDAF_BDGE_MISC_CBASE_H;
-		edaf |= FIELD_PREP(EDAF_BDGE_MISC_CBASE_H, cbase >> 32);
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_MISC);
-	}
+		rc = ofnode_read_u32(node, "cmd-read", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_READ);
 
-	rc = ofnode_read_u32(node, "cmd-read", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_READ);
+		rc = ofnode_read_u32(node, "cmd-write-enable", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_WRITE_EN);
 
-	rc = ofnode_read_u32(node, "cmd-write-enable", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_WRITE_EN);
+		rc = ofnode_read_u32(node, "cmd-write", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_WRITE);
 
-	rc = ofnode_read_u32(node, "cmd-write", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_WRITE);
+		rc = ofnode_read_u32(node, "cmd-read-status", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_READ_STS);
 
-	rc = ofnode_read_u32(node, "cmd-read-status", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_READ_STS);
+		rc = ofnode_read_u32(node, "cmd-erase-4k", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_4K);
 
-	rc = ofnode_read_u32(node, "cmd-erase-4k", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_4K);
+		rc = ofnode_read_u32(node, "cmd-erase-32k", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_32K);
 
-	rc = ofnode_read_u32(node, "cmd-erase-32k", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_32K);
+		rc = ofnode_read_u32(node, "cmd-erase-64k", &edaf);
+		if (!rc)
+			writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_64K);
+	}
+	writel(mbase, edaf_bridge_regs + EDAF_BDGE_MBASE);
 
-	rc = ofnode_read_u32(node, "cmd-erase-64k", &edaf);
-	if (!rc)
-		writel(edaf, edaf_bridge_regs + EDAF_BDGE_CMD_ERASE_64K);
+	edaf = readl(edaf_bridge_regs + EDAF_BDGE_MISC);
+	edaf &= ~EDAF_BDGE_MISC_MBASE_H;
+	edaf |= FIELD_PREP(EDAF_BDGE_MISC_MBASE_H, mbase >> 32);
+	writel(edaf, edaf_bridge_regs + EDAF_BDGE_MISC);
 
 	rc = ofnode_read_u64(node, "espi-base", &espibase);
 	if (!rc) {
-- 
2.34.1

